{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e7741a7",
   "metadata": {},
   "source": [
    "## Dec 2022\n",
    "## Explore PLR excel report, rank top species to notice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e5bb9586",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: openpyxl in /Users/huangz36/miniconda3/lib/python3.9/site-packages (3.0.10)\r\n",
      "Requirement already satisfied: et-xmlfile in /Users/huangz36/miniconda3/lib/python3.9/site-packages (from openpyxl) (1.1.0)\r\n"
     ]
    }
   ],
   "source": [
    "!pip install openpyxl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "88877021",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/huangz36/miniconda3/lib/python3.9/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.1\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as mtick\n",
    "plt.rcParams['figure.dpi'] = 600\n",
    "sns.set(style=\"whitegrid\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7e70624d",
   "metadata": {},
   "outputs": [],
   "source": [
    "oligo_nt_set = {'a', 'c', 'g', 't', '-', 'A', 'C','G', 'T'}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9360be74",
   "metadata": {},
   "source": [
    "## For each row [unique infectious disease strain id], calculate features and assign rankings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "52eecf24",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Foward/reverse primers\n",
    "def count_no_mismatches(oligo_string):\n",
    "    '''\n",
    "    Calculate no. mismatch and frac of mismatch in the oligo\n",
    "    '''\n",
    "    no_match_count = 0\n",
    "    oligo_string.replace(\".\", \"-\")\n",
    "    for letter in oligo_string:\n",
    "        if letter != \".\" or letter != \"-\":\n",
    "            no_match_count+=1\n",
    "    fraction_mismatch = round(no_match_count/len(oligo_string), 2)\n",
    "    return (no_match_count, fraction_mismatch)\n",
    "\n",
    "\n",
    "\n",
    "def three_prime_end_mismatch(oligo_string, no_3_prime_nts):\n",
    "    '''\n",
    "    Calculate no. of mismatch at 3' end of the oligo\n",
    "    '''\n",
    "    non_match_count = 0\n",
    "    three_prime = oligo_string[-no_3_prime_nts:]\n",
    "    for letter in three_prime:\n",
    "        if letter != \".\":\n",
    "            non_match_count+=1\n",
    "    return non_match_count\n",
    "\n",
    "def count_primer_relative_freq(oligo_sequence, df, oligo_column_name):\n",
    "    df_count_agg = df[['id', oligo_column_name]].groupby(oligo_column_name).count().reset_index()\n",
    "    freq_count = df_count_agg.loc[df_count_agg[oligo_column_name]==oligo_sequence, 'id'].values[0]\n",
    "    total_ids = df.shape[0]\n",
    "    freq_as_ratio = round(int(freq_count)/int(total_ids), 2)\n",
    "    return freq_count, freq_as_ratio\n",
    "\n",
    "def input_PLR_return_taxonIDs_need_notice(PLR_file, PLR_sheet_name, no_top_taxon_to_show, three_prime_distance):\n",
    "    '''\n",
    "    PLR_file: path to the PLR excel sheet\n",
    "    PLR_sheet_name: the name of the excel tab of \"*_report\", with oligo related information\n",
    "    no_top_taxon_to_show: no. of the most worth noticing taxons to show, that may not be amplified by assay\n",
    "    three_prime_distance: for forward and reverse primers, the n nucleotides to 3' end that needs mismatch check\n",
    "    '''\n",
    "    ## Read in dataframe, oligo names and column names\n",
    "    df = pd.read_excel(PLR_file, sheet_name=PLR_sheet_name)\n",
    "    oligo_header = df.columns.to_list()\n",
    "    complete_header = df.iloc[0].to_list()\n",
    "\n",
    "    ## Merge and organize new header\n",
    "    new_header = []\n",
    "    for column_no in range(len(oligo_header)):\n",
    "        if oligo_header[column_no].startswith(\"Unnamed\"):\n",
    "            new_col_name = complete_header[column_no]\n",
    "        elif set(complete_header[column_no]).issubset(oligo_nt_set):\n",
    "            new_col_name = oligo_header[column_no]\n",
    "        else:\n",
    "            new_col_name = str(str(oligo_header[column_no]).split(\".\")[0]) + \":\" + str(complete_header[column_no])\n",
    "        new_header.append(new_col_name)\n",
    "\n",
    "    ## Rename df with new_header\n",
    "    df = df[1:]\n",
    "    df.columns = new_header\n",
    "\n",
    "    ## For primer oligo sequence columns ending with (f) and (r), make mismatch related calculations\n",
    "    ## For primer oligo sequence, also count frequency and relative abundance of the haplotype\n",
    "    primer_sequence_columns = []\n",
    "    three_prime_rank_columns = []\n",
    "    no_mismatch_rank_columns = []\n",
    "    mismatch_ratio_rank_columns = []\n",
    "    haplotype_count_columns = []\n",
    "    for col_names in new_header:\n",
    "        #print(\"col_names\", col_names)\n",
    "        if col_names.endswith(\"(f)\") or col_names.endswith(\"(r)\"): # the oligo columns\n",
    "            three_prime_mismatch_name = str(col_names) + \"_3_prime_mismatch\"\n",
    "            mismatch_name = str(col_names) + \"_mismatch_count\"\n",
    "            mismatch_ratio_name = str(col_names) + \"_mismatch_ratio\"\n",
    "            haplotype_count_name = str(col_names) + \"_haplotype_count\"\n",
    "            haplotype_ratio_name = str(col_names) + \"_relative_abundance\"\n",
    "\n",
    "            df[three_prime_mismatch_name] = df[col_names].apply(three_prime_end_mismatch, args=(three_prime_distance, ))\n",
    "            df[[mismatch_name, mismatch_ratio_name]]= df[col_names].apply(count_no_mismatches).tolist()\n",
    "            df[[haplotype_count_name, haplotype_ratio_name]] = df[col_names].apply(count_primer_relative_freq, args=(df, col_names)).tolist()\n",
    "\n",
    "            three_prime_rank_columns.append(three_prime_mismatch_name)\n",
    "            no_mismatch_rank_columns.append(mismatch_name)\n",
    "            mismatch_ratio_rank_columns.append(mismatch_ratio_name)\n",
    "            haplotype_count_columns.append(haplotype_ratio_name)\n",
    "            primer_sequence_columns.append(col_names)\n",
    "            \n",
    "            columns_for_strain_ranking = three_prime_rank_columns + \\\n",
    "            no_mismatch_rank_columns + mismatch_ratio_rank_columns + haplotype_count_columns\n",
    "        ## Pending \n",
    "        ## need do develop for \"p\", once quencher and dye location obtained\n",
    "\n",
    "    ascending_list = [False] * len(columns_for_strain_ranking)\n",
    "    df_need_notice = df.sort_values(columns_for_strain_ranking, ascending = ascending_list)\n",
    "\n",
    "    ## Rank strain IDs\n",
    "    to_show_columns = ['Isolate_Id', 'Isolate_Name', 'Location', 'Host', 'Collection_Date', 'country']\n",
    "    to_show_extended = to_show_columns + primer_sequence_columns + haplotype_count_columns \\\n",
    "    + three_prime_rank_columns + mismatch_ratio_rank_columns\n",
    "    df_need_notice = df.sort_values(columns_for_strain_ranking, \\\n",
    "                                    ascending = ascending_list).head(no_top_taxon_to_show)[to_show_extended]\n",
    "    \n",
    "    return df_need_notice\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a58156fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PLR_file = '/Users/huangz36/Documents/Hackathon_2022/Galaxy2243.PLR.xlsx'\n",
    "#sheet_name = 'Yers_report'\n",
    "\n",
    "#PLR_file = '/Users/huangz36/Documents/Hackathon_2022/FluA_hackathon.xlsx'\n",
    "\n",
    "PLR_file = '/Users/huangz36/Documents/Hackathon_2022/flu_hackathon_1100.xlsx'\n",
    "sheet_name = 'FluA_report'\n",
    "\n",
    "\n",
    "df = pd.read_excel(PLR_file, sheet_name=sheet_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "4abdf445",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_df = input_PLR_return_taxonIDs_need_notice(PLR_file, sheet_name, 10, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9c27d423",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Isolate_Id</th>\n",
       "      <th>Isolate_Name</th>\n",
       "      <th>Location</th>\n",
       "      <th>Host</th>\n",
       "      <th>Collection_Date</th>\n",
       "      <th>country</th>\n",
       "      <th>FluA_assay(f)</th>\n",
       "      <th>FluA_assay(r)</th>\n",
       "      <th>FluA_assay(f)_relative_abundance</th>\n",
       "      <th>FluA_assay(r)_relative_abundance</th>\n",
       "      <th>FluA_assay(f)_3_prime_mismatch</th>\n",
       "      <th>FluA_assay(r)_3_prime_mismatch</th>\n",
       "      <th>FluA_assay(f)_mismatch_ratio</th>\n",
       "      <th>FluA_assay(r)_mismatch_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>640</th>\n",
       "      <td>A/Hawaii/46/2022|EPI_ISL_15928821||State_of_Ha...</td>\n",
       "      <td>EPI_ISL_15928821</td>\n",
       "      <td>A/Hawaii/46/2022</td>\n",
       "      <td>North America / United States / Hawaii</td>\n",
       "      <td>Human</td>\n",
       "      <td>2022-10-19 00:00:00</td>\n",
       "      <td>United States</td>\n",
       "      <td>...............T.-.......C</td>\n",
       "      <td>..................</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>A/Hawaii/45/2022|EPI_ISL_15928807||State_of_Ha...</td>\n",
       "      <td>EPI_ISL_15928807</td>\n",
       "      <td>A/Hawaii/45/2022</td>\n",
       "      <td>North America / United States / Hawaii</td>\n",
       "      <td>Human</td>\n",
       "      <td>2022-10-18 00:00:00</td>\n",
       "      <td>United States</td>\n",
       "      <td>...............T.-.......C</td>\n",
       "      <td>..................</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762</th>\n",
       "      <td>A/Colorado/36/2022|EPI_ISL_15928796||Colorado_...</td>\n",
       "      <td>EPI_ISL_15928796</td>\n",
       "      <td>A/Colorado/36/2022</td>\n",
       "      <td>North America / United States / Colorado</td>\n",
       "      <td>Human</td>\n",
       "      <td>2022-10-16 00:00:00</td>\n",
       "      <td>United States</td>\n",
       "      <td>...............T.-.......C</td>\n",
       "      <td>..................</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>786</th>\n",
       "      <td>A/Hawaii/43/2022|EPI_ISL_15928787||State_of_Ha...</td>\n",
       "      <td>EPI_ISL_15928787</td>\n",
       "      <td>A/Hawaii/43/2022</td>\n",
       "      <td>North America / United States / Hawaii</td>\n",
       "      <td>Human</td>\n",
       "      <td>2022-10-15 00:00:00</td>\n",
       "      <td>United States</td>\n",
       "      <td>...............T.-.......C</td>\n",
       "      <td>..................</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864</th>\n",
       "      <td>A/Hawaii/42/2022|EPI_ISL_15928830||State_of_Ha...</td>\n",
       "      <td>EPI_ISL_15928830</td>\n",
       "      <td>A/Hawaii/42/2022</td>\n",
       "      <td>North America / United States / Hawaii</td>\n",
       "      <td>Human</td>\n",
       "      <td>2022-10-13 00:00:00</td>\n",
       "      <td>United States</td>\n",
       "      <td>...............T.-.......C</td>\n",
       "      <td>..................</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>895</th>\n",
       "      <td>A/Hawaii/41/2022|EPI_ISL_15928818||State_of_Ha...</td>\n",
       "      <td>EPI_ISL_15928818</td>\n",
       "      <td>A/Hawaii/41/2022</td>\n",
       "      <td>North America / United States / Hawaii</td>\n",
       "      <td>Human</td>\n",
       "      <td>2022-10-12 00:00:00</td>\n",
       "      <td>United States</td>\n",
       "      <td>...............T.-.......C</td>\n",
       "      <td>..................</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>896</th>\n",
       "      <td>A/California/140/2022|EPI_ISL_15928786||Califo...</td>\n",
       "      <td>EPI_ISL_15928786</td>\n",
       "      <td>A/California/140/2022</td>\n",
       "      <td>North America / United States / California</td>\n",
       "      <td>Human</td>\n",
       "      <td>2022-10-12 00:00:00</td>\n",
       "      <td>United States</td>\n",
       "      <td>...............T.-.......C</td>\n",
       "      <td>..................</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1074</th>\n",
       "      <td>A/Iowa/26/2022|EPI_ISL_15715021||Iowa_State_Hy...</td>\n",
       "      <td>EPI_ISL_15715021</td>\n",
       "      <td>A/Iowa/26/2022</td>\n",
       "      <td>North America / United States / Iowa</td>\n",
       "      <td>Human</td>\n",
       "      <td>2022-10-07 00:00:00</td>\n",
       "      <td>United States</td>\n",
       "      <td>...............T.-.......C</td>\n",
       "      <td>..................</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1089</th>\n",
       "      <td>A/Iowa/26/2022|EPI_ISL_15928856||Iowa_State_Hy...</td>\n",
       "      <td>EPI_ISL_15928856</td>\n",
       "      <td>A/Iowa/26/2022</td>\n",
       "      <td>North America / United States / Iowa</td>\n",
       "      <td>Human</td>\n",
       "      <td>2022-10-07 00:00:00</td>\n",
       "      <td>United States</td>\n",
       "      <td>...............T.-.......C</td>\n",
       "      <td>..................</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494</th>\n",
       "      <td>A/Ibra/72211857/2022|EPI_ISL_15926932||Central...</td>\n",
       "      <td>EPI_ISL_15926932</td>\n",
       "      <td>A/Ibra/72211857/2022</td>\n",
       "      <td>Asia / Oman</td>\n",
       "      <td>Human</td>\n",
       "      <td>2022-10-23 00:00:00</td>\n",
       "      <td>Oman</td>\n",
       "      <td>...............T.-......G.</td>\n",
       "      <td>..................</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     id        Isolate_Id  \\\n",
       "640   A/Hawaii/46/2022|EPI_ISL_15928821||State_of_Ha...  EPI_ISL_15928821   \n",
       "705   A/Hawaii/45/2022|EPI_ISL_15928807||State_of_Ha...  EPI_ISL_15928807   \n",
       "762   A/Colorado/36/2022|EPI_ISL_15928796||Colorado_...  EPI_ISL_15928796   \n",
       "786   A/Hawaii/43/2022|EPI_ISL_15928787||State_of_Ha...  EPI_ISL_15928787   \n",
       "864   A/Hawaii/42/2022|EPI_ISL_15928830||State_of_Ha...  EPI_ISL_15928830   \n",
       "895   A/Hawaii/41/2022|EPI_ISL_15928818||State_of_Ha...  EPI_ISL_15928818   \n",
       "896   A/California/140/2022|EPI_ISL_15928786||Califo...  EPI_ISL_15928786   \n",
       "1074  A/Iowa/26/2022|EPI_ISL_15715021||Iowa_State_Hy...  EPI_ISL_15715021   \n",
       "1089  A/Iowa/26/2022|EPI_ISL_15928856||Iowa_State_Hy...  EPI_ISL_15928856   \n",
       "494   A/Ibra/72211857/2022|EPI_ISL_15926932||Central...  EPI_ISL_15926932   \n",
       "\n",
       "               Isolate_Name                                    Location  \\\n",
       "640        A/Hawaii/46/2022      North America / United States / Hawaii   \n",
       "705        A/Hawaii/45/2022      North America / United States / Hawaii   \n",
       "762      A/Colorado/36/2022    North America / United States / Colorado   \n",
       "786        A/Hawaii/43/2022      North America / United States / Hawaii   \n",
       "864        A/Hawaii/42/2022      North America / United States / Hawaii   \n",
       "895        A/Hawaii/41/2022      North America / United States / Hawaii   \n",
       "896   A/California/140/2022  North America / United States / California   \n",
       "1074         A/Iowa/26/2022        North America / United States / Iowa   \n",
       "1089         A/Iowa/26/2022        North America / United States / Iowa   \n",
       "494    A/Ibra/72211857/2022                                 Asia / Oman   \n",
       "\n",
       "       Host      Collection_Date        country               FluA_assay(f)  \\\n",
       "640   Human  2022-10-19 00:00:00  United States  ...............T.-.......C   \n",
       "705   Human  2022-10-18 00:00:00  United States  ...............T.-.......C   \n",
       "762   Human  2022-10-16 00:00:00  United States  ...............T.-.......C   \n",
       "786   Human  2022-10-15 00:00:00  United States  ...............T.-.......C   \n",
       "864   Human  2022-10-13 00:00:00  United States  ...............T.-.......C   \n",
       "895   Human  2022-10-12 00:00:00  United States  ...............T.-.......C   \n",
       "896   Human  2022-10-12 00:00:00  United States  ...............T.-.......C   \n",
       "1074  Human  2022-10-07 00:00:00  United States  ...............T.-.......C   \n",
       "1089  Human  2022-10-07 00:00:00  United States  ...............T.-.......C   \n",
       "494   Human  2022-10-23 00:00:00           Oman  ...............T.-......G.   \n",
       "\n",
       "           FluA_assay(r)  FluA_assay(f)_relative_abundance  \\\n",
       "640   ..................                              0.01   \n",
       "705   ..................                              0.01   \n",
       "762   ..................                              0.01   \n",
       "786   ..................                              0.01   \n",
       "864   ..................                              0.01   \n",
       "895   ..................                              0.01   \n",
       "896   ..................                              0.01   \n",
       "1074  ..................                              0.01   \n",
       "1089  ..................                              0.01   \n",
       "494   ..................                              0.00   \n",
       "\n",
       "      FluA_assay(r)_relative_abundance  FluA_assay(f)_3_prime_mismatch  \\\n",
       "640                               0.71                               1   \n",
       "705                               0.71                               1   \n",
       "762                               0.71                               1   \n",
       "786                               0.71                               1   \n",
       "864                               0.71                               1   \n",
       "895                               0.71                               1   \n",
       "896                               0.71                               1   \n",
       "1074                              0.71                               1   \n",
       "1089                              0.71                               1   \n",
       "494                               0.71                               1   \n",
       "\n",
       "      FluA_assay(r)_3_prime_mismatch  FluA_assay(f)_mismatch_ratio  \\\n",
       "640                                0                           1.0   \n",
       "705                                0                           1.0   \n",
       "762                                0                           1.0   \n",
       "786                                0                           1.0   \n",
       "864                                0                           1.0   \n",
       "895                                0                           1.0   \n",
       "896                                0                           1.0   \n",
       "1074                               0                           1.0   \n",
       "1089                               0                           1.0   \n",
       "494                                0                           1.0   \n",
       "\n",
       "      FluA_assay(r)_mismatch_ratio  \n",
       "640                            1.0  \n",
       "705                            1.0  \n",
       "762                            1.0  \n",
       "786                            1.0  \n",
       "864                            1.0  \n",
       "895                            1.0  \n",
       "896                            1.0  \n",
       "1074                           1.0  \n",
       "1089                           1.0  \n",
       "494                            1.0  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "003cdaf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_df = input_PLR_return_taxonIDs_need_notice(PLR_file, sheet_name, 10, 4)\n",
    "output_df.to_csv(\"/Users/huangz36/Documents/Hackathon_2022/example_taxon_at_risk.csv\", index=False, header=True)\n",
    "df.to_csv(\"/Users/huangz36/Documents/Hackathon_2022/processed_PLR_risk.csv\", index=False, header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
